ouroboros-consensus-test/bench-mempool/Bench/Consensus/Mempool/Scenario.hs

```diff
@@ -0,0 +1,241 @@
+  , scChangelogBlocks :: ![TestBlock]
+    -- | Mempool commands to run.
+  , scMempoolCommands :: ![MempoolCmd TestBlock]
+  }
+
+-- | Convert a 'Scenario' into mempool parameters and a benchmark workload
+-- (commands).
+fromScenario ::
+     LedgerDbCfg (LedgerState TestBlock)
+  -> BackingStoreSelector m
+  -> Scenario
+  -> (InitialMempoolAndModelParams m TestBlock, [MempoolCmd TestBlock])
+fromScenario ldbCfg bss sc =
+  (mkParams
+    (scBackingState sc)
+    (scChangelogBlocks sc)
+    ldbCfg
+    bss
+  , scMempoolCommands sc
+  )
+
+{-------------------------------------------------------------------------------
+  Scenario builder
+-------------------------------------------------------------------------------}
+
+-- | A monadic interface to constructing benchmark scenarios.
+newtype ScBuilder a = ScBuilder { runScBuilder :: State St a }
+  deriving newtype (Functor, Applicative, Monad, MonadState St)
+
+-- | Internal state for the 'ScenarioBuilder' monad.
+data St = St {
+    -- | A new, unique 'Token'. Reading this field should always be followed by
+    -- an increment.
+    stNextToken :: !Token
+    -- | The scenario being built.
+  , stScenario  :: !Scenario
+  }
+
+initialSt :: St
+initialSt = St 0 (Scenario (ledgerStateFromTokens []) [] [])
+
+execScBuilder :: ScBuilder a -> St -> St
+execScBuilder = execState . runScBuilder
+
+build :: ScBuilder () -> Scenario
+build m = stScenario (execScBuilder m initialSt)
+
+{-------------------------------------------------------------------------------
+  Scenario builder: 'MonadState' extras
+-------------------------------------------------------------------------------}
+
+modifyScenario :: MonadState St m => (Scenario -> Scenario) -> m ()
+modifyScenario f = modify (\st -> st {
+      stScenario = f $ stScenario st
+    })
+
+putBackingState :: MonadState St m => LedgerState TestBlock ValuesMK -> m ()
+putBackingState lst = modifyScenario $ \sc ->
+    precondition sc $ sc {scBackingState = lst }
+  where
+    precondition sc = assert (putBackingStateOnlyOnce sc)
+    putBackingStateOnlyOnce sc = scBackingState sc == ledgerStateFromTokens []
+
+putChangelogBlocks :: MonadState St m => [TestBlock] -> m ()
+putChangelogBlocks blks = modifyScenario $ \sc ->
+    precondition sc $ sc { scChangelogBlocks = blks }
+  where
+    precondition sc = assert (putChangelogBlocksOnlyOnce sc)
+    putChangelogBlocksOnlyOnce sc = null $ scChangelogBlocks sc
+
+putMempoolCommands :: MonadState St m => [MempoolCmd TestBlock] -> m ()
+putMempoolCommands cmds = modifyScenario $ \sc ->
+    precondition sc $ sc { scMempoolCommands = cmds }
+  where
+    precondition sc = assert (putMempoolCommandsOnlyOnce sc)
+    putMempoolCommandsOnlyOnce sc = null $ scMempoolCommands sc
+
+{-------------------------------------------------------------------------------
+  Scenario builder: main API
+-------------------------------------------------------------------------------}
+
```

Comment:
Should we comment that this outputs `n` transactions that consume nothing and produce 1 token, and that the backing store state is set to contain exactly the tokens produced by this function?

Reply:
If so, perhaps we should add comments to the functions below.

Reply:
Done